cfgov/core/tests/test_views.py

```diff
@@ -1,221 +1,257 @@
+from django.test import RequestFactory, TestCase
 from django.core.urlresolvers import reverse
 from django.http import QueryDict
 
 from django.contrib.messages.storage.cookie import CookieStorage
 from django.contrib.messages import SUCCESS
 from requests_toolbelt.multipart.encoder import MultipartEncoder
-from core.views import submit_comment
+
+from core.utils import extract_answers_from_request
+from core.views import govdelivery_subscribe, regsgov_comment, submit_comment
+
+
+django.setup()
 
 
 class GovDeliverySubscribeTest(TestCase):
-    """
-    This TestCase mocks the following:
-    authenticated_session - removes the external govdelivery call
-    set_subscriber_topics - removes another external govdelivery call
-    assertEquals(urlparse(url).path, reverse(route_name)) was used over
-    assertRedirects because the former doesn't require the templates to exist
-    """
+    def setUp(self):
+        self.factory = RequestFactory()
+
+    def post(self, post, ajax=False):
+        kwargs = {'HTTP_X_REQUESTED_WITH': 'XMLHttpRequest'} if ajax else {}
+        request = self.factory.post(reverse('govdelivery'), post, **kwargs)
+        return govdelivery_subscribe(request)
+
+    def assertRedirect(self, response, redirect):
+        self.assertEqual(
+            (response['Location'], response.status_code),
+            (reverse(redirect), 302)
+        )
+
+    def assertRedirectSuccess(self, response):
+        self.assertRedirect(response, 'govdelivery:success')
+
+    def assertRedirectUserError(self, response):
+        self.assertRedirect(response, 'govdelivery:user_error')
+
+    def assertRedirectServerError(self, response):
+        self.assertRedirect(response, 'govdelivery:server_error')
+
+    def assertJSON(self, response, result):
+        self.assertEqual(
+            response.content.decode('utf-8'),
+            json.dumps({'result': result})
+        )
+
+    def assertJSONSuccess(self, response):
+        return self.assertJSON(response, 'pass')
+
+    def assertJSONError(self, response):
+        return self.assertJSON(response, 'fail')
+
+    def check_post(self, post, response_check, ajax=False):
+        response_check(self.post(post, ajax=ajax))
+
+    def mock_govdelivery(self, status_code=200):
+        gd = Mock(set_subscriber_topics=Mock(
+            return_value=Mock(status_code=status_code)
+        ))
+
+        patcher = patch('core.views.GovDelivery', return_value=gd)
+        patched = patcher.start()
+        self.addCleanup(patcher.stop)
+
+        return gd
+
+    def check_subscribe(self, response_check, ajax=False, status_code=200,
+                        include_answers=False):
+        post = {
+            'code': 'FAKE_CODE',
+            'email': 'fake@example.com',
+        }
+
+        answers = [
+            ('batman', 'robin'),
+            ('hello', 'goodbye'),
+        ]
+
+        if include_answers:
+            post.update({'questionid_' + q: a for q, a in answers})
```

Comment:
This code remains quite hacky although hopefully an improvement on what was there before. It'd be better to refactor the underlying govdelivery code to allow for nicer testing of individual methods. For this PR I tried to focus solely on maintaining existing tests while decoupling from the DB.